File: android-plugin-src/java/com/treerise/app/usage/SyncWorker.java
package com.treerise.app.usage;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.json.JSONArray;
import org.json.JSONObject;

import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Map;

public class SyncWorker extends Worker {

    public SyncWorker(@NonNull Context context, @NonNull WorkerParameters params) {
        super(context, params);
    }

    @NonNull
    @Override
    public Result doWork() {
        String endpoint = getInputData().getString("endpoint");
        String bearer = getInputData().getString("bearer");
        if (endpoint == null || bearer == null) return Result.failure();
        if (!UsageStatsPlugin.isUsageAccessGranted(getApplicationContext())) return Result.success();

        Map<String, Long> usage = UsageStatsPlugin.readTodayUsage(getApplicationContext());
        JSONArray arr = new JSONArray();
        try {
            for (Map.Entry<String, Long> e : usage.entrySet()) {
                JSONObject o = new JSONObject();
                o.put("packageName", e.getKey());
                o.put("minutes", e.getValue());
                arr.put(o);
            }
            JSONObject body = new JSONObject();
            body.put("entries", arr);

            URL url = new URL(endpoint);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("POST");
            conn.setDoOutput(true);
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setRequestProperty("Authorization", "Bearer " + bearer);
            conn.setConnectTimeout(15000);
            conn.setReadTimeout(15000);
            conn.getOutputStream().write(body.toString().getBytes());
            conn.getOutputStream().close();
            int code = conn.getResponseCode();
            conn.disconnect();
            return (code >= 200 && code < 300) ? Result.success() : Result.retry();
        } catch (Exception e) {
            return Result.retry();
        }
    }
}

File: android-plugin-src/java/com/treerise/app/usage/SyncWorker.kt
package com.treerise.app.usage

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class SyncWorker(appContext: Context, params: WorkerParameters) :
    CoroutineWorker(appContext, params) {

    override suspend fun doWork(): Result {
        val endpoint = inputData.getString("endpoint") ?: return Result.failure()
        val bearer = inputData.getString("bearer") ?: return Result.failure()
        if (!UsageStatsPlugin.isUsageAccessGranted(applicationContext)) return Result.success()

        val usage = UsageStatsPlugin.readTodayUsage(applicationContext)
        val arr = JSONArray()
        for ((pkg, mins) in usage) {
            val o = JSONObject()
            o.put("packageName", pkg)
            o.put("minutes", mins)
            arr.put(o)
        }
        val body = JSONObject()
        body.put("entries", arr)

        return try {
            val url = URL(endpoint)
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.doOutput = true
            conn.setRequestProperty("Content-Type", "application/json")
            conn.setRequestProperty("Authorization", "Bearer $bearer")
            conn.connectTimeout = 15000
            conn.readTimeout = 15000
            conn.outputStream.use { it.write(body.toString().toByteArray()) }
            val code = conn.responseCode
            conn.disconnect()
            if (code in 200..299) Result.success() else Result.retry()
        } catch (e: Exception) {
            Result.retry()
        }
    }
}
